### PR TITLE
[profiles] introduce semantic convention

### PR DESCRIPTION
**Pull request by florianl**
## Changes

With https://github.com/open-telemetry/opentelemetry-proto/pull/534 a new OTel signal was introduced.

This change introduced semantic conventions specific to the OTel Profiling signal.

FYI: @open-telemetry/profiling-maintainers  @aalexand 

## Merge requirement checklist

* [x] [CONTRIBUTING.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md) guidelines followed.
* [x] Change log entry added, according to the guidelines in [When to add a changelog entry](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md#when-to-add-a-changelog-entry).
  * If your PR does not need a change log, start the PR title with `[chore]`
* [ ] [schema-next.yaml](https://github.com/open-telemetry/semantic-conventions/blob/main/schema-next.yaml) updated with changes to existing conventions.

### CODE REVIEW

File: .chloggen/profiles-convention.yaml
Comment (trisch-me) [inline .chloggen/profiles-convention.yaml:7]:
```suggestion
change_type: new_component
```

File: .chloggen/profiles-convention.yaml
Comment (trisch-me) [inline .chloggen/profiles-convention.yaml:10]:
```suggestion
component: profile
```

**Review by petethepig — APPROVED** [review]

File: model/registry/profile.yaml
Comment (trask) [inline model/registry/profile.yaml:28]: @open-telemetry/specs-semconv-approvers thoughts?

* `trace.id` is in line with the direction we seem to be heading: https://github.com/open-telemetry/semantic-conventions/issues/1118#issuecomment-2159136053
* `trace_id` matches the logging spec's [Trace Context in Legacy Formats](
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/README.md#trace-context-in-legacy-formats) (see https://github.com/open-telemetry/opentelemetry-specification/issues/3303)

File: model/registry/profile.yaml
Comment (trask) [inline model/registry/profile.yaml:101]: might be nice if this aligned with the namespace in #1066 (either changing that to `v8` or this to `v8js`)

cc @maryliag

Comment (florianl) [reply model/registry/profile.yaml:101]: `v8js` sounds also good :+1: `v8` was chosen as this is the value that has been used by Elastic in the past. But going forward and aligning with other components, `v8js` sounds right.

File: model/registry/profile.yaml
Comment (AlexanderWert) [reply model/registry/profile.yaml:28]: Today, we have `trace_id` in legacy formats for logs and the corresponding OTLP-native field.

So, `trace.id` would be a new (yet another representation) of the same thing. That's why I'm slightly more in favor of  `trace_id`. Also, `trace.id` would make sense if we expect more attributes under the `trace` namespace, which I think is not the case here.

File: docs/general/profiles.md
Comment (lmolkova) [inline docs/general/profiles.md:46]: It seems profiles will have dedicated properties for span-id and trace-id https://github.com/open-telemetry/opentelemetry-proto/blob/40b3c1b746767cbc13c2e39da3eaf1a23e54ffdd/opentelemetry/proto/profiles/v1experimental/pprofextended.proto#L94

What would be the use-case for `profile.link.*`? what they will be stamped on? spans? logs? profiles themselves?

File: docs/general/profiles.md
Comment (lmolkova) [inline docs/general/profiles.md:108]: semconv allows [a-z0-9_\.] chars https://github.com/open-telemetry/semantic-conventions/issues/1124, we also have no tooling support to reserve the namespace at the moment.

But also, who/how/when will use these attributes? what would they be stamped on?

Comment (lmolkova) [reply docs/general/profiles.md:108]: created https://github.com/open-telemetry/semantic-conventions/issues/1197 to track reserving namespace

File: docs/attributes-registry/profile.md
Comment (lmolkova) [inline docs/attributes-registry/profile.md:19]: is there some documentation on this? could it be referenced? So someone not familiar with profiling could go and learn what it means?

File: docs/general/profiles.md
Comment (florianl) [reply docs/general/profiles.md:46]: One task of the OTel Profiling SIG is to have some compatibility between OTel profiling and pprof (tool + protocol) - see also  https://github.com/open-telemetry/opentelemetry-proto/issues/567.

The semantic convention for OTel profiles allows to address this topic. And so the idea is to define the connection between profiles and spanes/traces within the semantic convention and remove `message Link` from the experimental OTel profiling protocol. This allows interoperability between OTel profiling and pprof.

File: docs/general/profiles.md
Comment (florianl) [reply docs/general/profiles.md:108]: Thanks for pointing out the allowed array of characters`[a-z0-9_.]` in Attributes. I wasn't aware of that, when writing this. Is there another way to express colons?

Compatibility between OTel profiling and pprof (tool + protocol) is important to the OTel profiling SIG. pprof is a well established tool and protocol in the industry. And there are special features, like `pprof::*` that pprof (tool) interprets in a special way. To avoid conflicts and guarantee compatibility, `pprof::*` should be avoided.

I'm aware that this is a special case and to my knowledge, semantic convention does not cover such special cases at the moment. This is also the reason, this section is written manually here and not automatically generated like other attributes.

File: docs/attributes-registry/profile.md
Comment (florianl) [reply docs/attributes-registry/profile.md:19]: Would it help to link here the GNU build system article from wikipedia?  https://wikipedia.org/wiki/GNU_build_system

Comment (tigrannajaryan) [reply docs/attributes-registry/profile.md:19]: We typically don't use signal-specific prefixes for attribute names, unless the attribute is only meaningful just for that one signal. Is `build_id` strictly a profiling concept? If it is an identifier of a build I think it may possibly be used in other signals (e.g. log an event and include information about the build of the executable that logged it). In that case we should not use `profile.` namespace and should put this attribute in an appropriate top-level namespace (or create one if none exists).
The same comment applies to other attributes introduced.

File: model/registry/profile.yaml
Comment (tigrannajaryan) [reply model/registry/profile.yaml:28]: Whatever we choose let's make sure it is consistent with what we do for logs. For now that means using `trace_id`/`span_id`.

File: docs/general/profiles.md
Comment (jsuereth) [reply docs/general/profiles.md:46]: These compatibility layer things MUST use `otel` namespace and are generally defined elsewhere.

See: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/semantic-conventions.md#reserved-namespace

File: docs/attributes-registry/profile.md
Comment (aalexand) [inline docs/attributes-registry/profile.md:20]: For the record, it looks like Go build ID is not handled correctly in Go's runtime/pprof and pprof CLI itself. I filed https://github.com/golang/go/issues/68652.

@felixge FYI

File: model/registry/profile.yaml
Comment (florianl) [reply model/registry/profile.yaml]: Removed the attributes around span/trace ID as there is a dedicated field in the OTel profiling signal.

File: docs/general/profiles.md
Comment (florianl) [reply docs/general/profiles.md]: This compatability section around google/pprof got removed.

File: docs/attributes-registry/profile.md
Comment (florianl) [reply docs/attributes-registry/profile.md:18]: @tigrannajaryan I do need some advice on resolving a conflict.

With `process.executable` there exists an attribute group, where `process.executable.build_id` could be another field. But for the Profiling use case, there is a conflict, that a single process can have multiple executables - e.g. a stack trace of a Python script can have libc, libpython, kernel and interpreted python frames (and maybe even more). The former three, libc, libpython and kernel, will have their dedicated `build_id`. So a 1:1 mapping of process to executable is not possible.
How could the 1:1 mapping become a 1:n mapping where one process can have multiple executables in its stack?  

FYI: @christos68k @AlexanderWert

Comment (florianl) [reply docs/attributes-registry/profile.md]: Removed the build_id from this proposed change with https://github.com/open-telemetry/semantic-conventions/pull/1188/commits/2ae1ac9c7156438f3698e9b0fb1f5ca182157926 and will open a new issue/PR around it as a follow up.

File: model/registry/profile.yaml
Comment (florianl) [reply model/registry/profile.yaml]: build_id got removed from this PR with https://github.com/open-telemetry/semantic-conventions/pull/1188/commits/2ae1ac9c7156438f3698e9b0fb1f5ca182157926.

**Comment by florianl** [comment]
@open-telemetry/profiling-maintainers the PR got updated to only contain the frame types.

**Review by christos68k — APPROVED** [review]

File: model/registry/profile.yaml
Comment (athre0z) [inline model/registry/profile.yaml:63]: This one is a bit of an outlier: all the other frame types here indicate a *language*, not an *implementation*. For example, we have `python` instead of `CPython`, `dotnet` instead of `.NET core`, and `jvm` instead of `hotspot`. V8, on the other hand, is a specific runtime. Should we generalize this to "js"?

Comment (florianl) [reply model/registry/profile.yaml:63]: The change from `v8` to `v8js` originates from a request in https://github.com/open-telemetry/semantic-conventions/pull/1188#discussion_r1658010097 and as https://github.com/open-telemetry/semantic-conventions/pull/1066 did land first, this proposed change did adapt.

Comment (athre0z) [reply model/registry/profile.yaml:63]: `v8js` makes sense in the context of what #1066 added because the attributes are actually runtime specific, but this isn't really the case here. 

I can't speak for @trask, but I suspect that his comment was based on the fact that the field already had the runtime-specific name "v8" in your draft. So if we went with that, it'd certainly make sense to go with the already established term `v8js`.

I think both approaches, reporting the general language OR the specific runtime would be raesonable for this field, and I don't have a strong preference. However, we should be *consistent* with that, and not mix it up at random.

**Review by brancz — APPROVED** [review]

File: model/registry/profile.yaml
Comment (joaopgrassi) [inline model/registry/profile.yaml:18]:
```suggestion
                [.NET](https://wikipedia.org/wiki/.NET)
```

File: model/registry/profile.yaml
Comment (joaopgrassi) [reply model/registry/profile.yaml:63]: .NET is not the language but the runtime/framework. You can program in C#, F# and Visual Basic with .NET. Also there's no .NET Core anymore, it's all only .NET.

Comment (athre0z) [reply model/registry/profile.yaml:63]: Fair point. The same is also true for JVM: it can also run e.g. Scala, so it's a runtime as well. For most languages in the list, the name of the reference implementation runtime is equal to the language itself. So I think if we'd just rename `Python` to `CPython`, it may already be consistent then.

Comment (florianl) [reply model/registry/profile.yaml:63]: https://github.com/open-telemetry/semantic-conventions/pull/1188/commits/608fb5c84773a3c7d060ebafd351677336e0839b

**Review by AlexanderWert — APPROVED** [review]

**Review by athre0z — APPROVED** [review]

File: model/registry/profile.yaml
Comment (felixge) [inline model/registry/profile.yaml:36]: Are these 4 languages grouped together on purpose, or because it's difficult for the ebpf profiler to distinguish them? Did we write down any use cases for these frame types somewhere? IIRC from the SIG discussions, the use case is to visually highlight different languages when visualizing the data (e.g. with different colors in a flame graph). If that's the case, distinguishing between C, C++, Go and Rust would probably be useful?

Apologies if this has been discussed already.

**Review by felixge — APPROVED** [review]

File: model/registry/profile.yaml
Comment (florianl) [reply model/registry/profile.yaml:36]: These four named languages are just an example for `native` frames and the list is not exhaustive.
From a pure stack unwinding point of view, I think it might be hardly possible to distinguish them. But by examining the backing executable/shared library, it should be possible. At the moment, this additional inspection is not done and needs more research.

Overall, `native` is just a general grouping and I think there should be no issue being more specific in the future and adding more specific languages, like `rust`.

Comment (felixge) [reply model/registry/profile.yaml:36]: Sounds good. Thanks!

**Comment by AlexanderWert** [comment]
@open-telemetry/specs-semconv-approvers Can we get more general approvals on this one please?
We already have a bunch of the profiling SIG's approvals, just missing an additional general semconv SIG approval.

File: model/profiles/frame.yaml
Comment (joaopgrassi) [inline model/profiles/frame.yaml:8]: Is it really opt in here? Wouldn't we always want to know the frame type? 🤔

File: docs/general/profiles.md
Comment (joaopgrassi) [inline docs/general/profiles.md:44]: You need to re-generate the table with `make fix` after changing to `cpython`

Comment (joaopgrassi) [reply docs/general/profiles.md:44]: @lmolkova shouldn't some check fail due to outdated markdown? The id of one of the enum names changed, but it is not reflected in the markdown here 🤔

Comment (florianl) [reply docs/general/profiles.md]: Did run `make fix` which fixed the name issue - see https://github.com/open-telemetry/semantic-conventions/pull/1188/commits/b6aed34588cc8313eb789884f87d7acc54a3f7c4. Not sure, why some commands did generate the correct output, like in `docs/attributes-registry/profile.md`, but this file did not get updated.

File: model/profiles/frame.yaml
Comment (florianl) [reply model/profiles/frame.yaml:8]: The understanding and agreement is, that `profile.frame.type` is optional, as it is "just" required for visualization of the data. Data without `profile.frame.type` can still be processed and provides its value.

Comment (joaopgrassi) [reply model/profiles/frame.yaml:8]: I see. Maybe then `recommended` is better here? `opt-in` requires the explicit opt-in from the user via configuration/sdk setup whereas recommended is "on" by default, but can be not added if there's good reason not to, like if it's expensive to collect and etc https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/attribute-requirement-level.md#recommended

File: docs/general/profiles.md
Comment (joaopgrassi) [reply docs/general/profiles.md]: Most likely you ran only the command to update the registry. They are separated and `make fix` executes both :)

**Review by joaopgrassi — APPROVED** [review]
Looks good to me, apart from the `opt-in` requirement, but I don't have a strong argument against it, so I will defer the decision to the profiling sig.

File: model/profiles/frame.yaml
Comment (florianl) [reply model/profiles/frame.yaml]: Updated the level to `recommended` with https://github.com/open-telemetry/semantic-conventions/pull/1188/commits/94bb7ed8f9f3288bffa02744058071c3d36ca4af.

**Pull request merged**
